client/consensus/babe/src/rpc.rs

```diff
@@ -0,0 +1,235 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! rpc api for babe.
+
+use std::fmt;
+use futures::{
+	prelude::*,
+	executor::ThreadPool,
+	channel::oneshot,
+	future,
+};
+use jsonrpc_core::BoxFuture;
+use jsonrpc_derive::rpc;
+use sp_consensus_babe::{AuthorityId, Epoch, AuthorityIndex, BabePreDigest};
+use crate::{SharedEpochChanges, authorship, epoch_changes::descendent_query, Config};
+use sc_keystore::KeyStorePtr;
+use std::sync::Arc;
+use sp_core::{crypto::Pair, Bytes};
+use sp_runtime::traits::{Block as BlockT, Header as _};
+use sp_consensus::{SelectChain, Error as ConsensusError};
+use sp_consensus_babe::BabeApi;
+use sp_blockchain::{HeaderBackend, HeaderMetadata, Error as BlockChainError};
+use serde::{Deserialize, Serialize};
+use sp_api::{ProvideRuntimeApi, BlockId};
+
+/// Provides rpc methods for interacting with Babe
+#[rpc(server)]
+pub trait Babe {
+	/// query slot authorship info
+	#[rpc(name = "babe_epochAuthorship")]
+	fn epoch_authorship(&self) -> BoxFuture<Vec<SlotAuthorship>>;
+}
+
+/// RPC handler for Babe
+/// provides `babe_epochAuthorship` method for querying slot authorship data.
+///
+/// Uses a background thread to calculate epoch_authorship data.
+pub struct BabeRPC<B: BlockT, C> {
+	/// shared refernce to the client.
+	client: Arc<C>,
+	/// shared reference to EpochChanges
+	shared_epoch_changes: SharedEpochChanges<B>,
+	/// shared reference to the Keystore
+	keystore: KeyStorePtr,
+	/// config (actually holds the slot duration)
+	babe_config: Config,
+	/// threadpool for spwaning cpu bound tasks.
+	threadpool: ThreadPool,
+}
+
+impl<B: BlockT, C> BabeRPC<B, C> {
+	/// creates a new instance of the BabeRpc handler.
+	pub fn new(
+		client: Arc<C>,
+		shared_epoch_changes: SharedEpochChanges<B>,
+		keystore: KeyStorePtr,
+		babe_config: Config
+	) -> Self {
+		let threadpool = ThreadPool::builder()
+			// single thread is fine.
+			.pool_size(1)
+			.create()
+			.unwrap();
+		Self {
+			client,
+			shared_epoch_changes,
+			keystore,
+			babe_config,
+			threadpool,
+		}
+	}
+}
+
+impl<B, C> Babe for BabeRPC<B, C>
+	where
+		B: BlockT,
+		C: ProvideRuntimeApi<B> + SelectChain<B> + HeaderBackend<B>
+		+ HeaderMetadata<B, Error=BlockChainError> + 'static,
+		C::Api: BabeApi<B>,
+		<<C as ProvideRuntimeApi<B>>::Api as sp_api::ApiErrorExt>::Error: fmt::Debug
+{
+	fn epoch_authorship(&self) -> BoxFuture<Vec<SlotAuthorship>> {
+		let (
+			babe_config,
+			keystore,
+			shared_epoch,
+			client,
+		) = (
+			self.babe_config.clone(),
+			self.keystore.clone(),
+			self.shared_epoch_changes.clone(),
+			self.client.clone(),
+		);
+		let (tx, rx) = oneshot::channel();
+
+		let future = async move {
```

Comment:
Why do we need to spawn this extra future?

Reply:
So the thing is calculating epoch authorship data requires cpu time, doing this work on jsonrpc threads (by returning the future directly) would prevent other rpc requests from happening, so instead of blocking jsonrpc threads by returning a cpu bound future, we uses a background thread and a one shot.

Reply:
I don't think it's a good idea either. The RPC servers should handle such calls just fine, we are doing plenty of blocking / runtime calls there already. We can probably figure out a global solution to this problem, rpc-specific optimizations like this will just produce a lot of noise in the code.